Sort reportActions by created instead of sequenceNumber

roryabraham opened the pull request:
### Details
Start sorting reportActions by their `created` timestamp, falling back on `actionname` / type as a secondary comparator in the event of same-millisecond reportActions, and finally falling back on `reportActionID` as a last comparator in the event of same-millisecond reportActions AND same `actionname`.

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/240894

### Tests / QA Steps
1. Open a chat
1. Scroll all the way back to the beginning. Verify that you see the report welcome message at the start of the chat.
1. Send a few messages in the chat.
1. Verify that the new chats appear in order at the end of the chat.
1. Turn off your internet
1. Send a few more messages in the chat.
1. Those chats should appear as pending while you're offline, but they should appear in the correct order at the end of the chat.
1. Come back online.
1. The chats should become finalized and no longer be pending.
1. Delete a message, verify that it disappears.
1. Go offline
1. Delete a message, verify that it appears as "pending deletion" (aka strikethrough)
1. Come back on line. Verify that the message you deleted while offline disappears.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).

- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The reviewer will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### Screenshots

#### Web
<img width="1792" alt="image" src="https://user-images.githubusercontent.com/47436092/204670771-09f1fc81-7b46-41f2-b722-36f3eecc2dfe.png">
<img width="1792" alt="image" src="https://user-images.githubusercontent.com/47436092/204670810-ef8d6897-9783-4ffa-b6f0-1a161a9b363e.png">
<img width="1792" alt="image" src="https://user-images.githubusercontent.com/47436092/204670874-86f026c8-a8b7-4c6a-9a1b-3dfa835e4976.png">
<img width="1792" alt="image" src="https://user-images.githubusercontent.com/47436092/204670911-2c2ed4ea-6114-41bf-9631-ed7b1148bce3.png">

#### Mobile Web - Chrome
<img src="https://user-images.githubusercontent.com/47436092/204671676-5617c38a-4fc0-4a35-9847-53ecaa4ade41.png" alt="" width="360px" />

#### Mobile Web - Safari
<img src="https://user-images.githubusercontent.com/47436092/204671234-c345db3e-be0f-4671-9422-55f67c73de65.PNG" alt="" width="360px" />
<img src="https://user-images.githubusercontent.com/47436092/204671282-00025584-6ed7-48d0-89e6-5d1ac35635ea.jpeg" alt="" width="360px" />

#### Desktop
<img width="1224" alt="image" src="https://user-images.githubusercontent.com/47436092/204672712-714777be-eda4-4cf7-a7d0-0febd0198751.png">
<img width="1228" alt="image" src="https://user-images.githubusercontent.com/47436092/204672759-0ed723e5-ffd9-48e0-8a97-22edf82a2671.png">
<img width="1227" alt="image" src="https://user-images.githubusercontent.com/47436092/204672784-84341ece-9fc1-452c-8289-8d372901ca85.png">

#### iOS
<img src="https://user-images.githubusercontent.com/47436092/204674559-fa0ebce0-48ab-45ca-baa0-db4b158e8781.jpeg" alt="" width="360px" />

#### Android
![Screenshot_20221129-161646](https://user-images.githubusercontent.com/47436092/204676884-a7a1e592-8105-436b-83b0-6a7389d55cde.png)

roryabraham committed:
Implement sortReportActions function

roryabraham committed:
Move sortReportActions to the top of ReportActionsUtils

roryabraham committed:
Preemptively re-implement getMostRecentIOUReportActionID

roryabraham committed:
Merge branch 'Rory-MostRecentIOUReportActionID' into Rory-SortReportActionsByCreated

# Conflicts:
#	src/libs/ReportActionsUtils.js

roryabraham committed:
Merge branch 'Rory-SequenceNumReportActionID' into Rory-SortReportActionsByCreated

roryabraham committed:
use sortReportActions to sort view

roryabraham committed:
Fix typo and some outdated item.action references

roryabraham committed:
Fix typo and add FIXME for created

roryabraham committed:
Merge branch 'main' into Rory-SortReportActionsByCreated

roryabraham committed:
Merge branch 'main' into Rory-SortReportActionsByCreated

# Conflicts:
#	src/libs/ReportActionsUtils.js
#	src/pages/home/report/ReportActionsList.js

roryabraham committed:
Remove outdated FIXME

roryabraham committed:
Update a few new usages of sortBy with sequenceNumber

roryabraham committed:
Fix a few bugs from bad merge

roryabraham committed:
Reverse sorting order in ReportActionsView for inverted FlatList

roryabraham committed:
Implement faster inverted sorting order

roryabraham committed:
Fix variable naming

roryabraham committed:
Remove outdated comment

sobitneupane reviewed (commented):
Didn't get the expected result. Got following screens while scrolling through previous chats.
#### Safari:
<img width="1418" alt="Screenshot 2022-11-23 at 14 32 07" src="https://user-images.githubusercontent.com/25876548/203504249-4278c275-17e7-4162-901e-71a999856581.png">

#### Chrome:
<img width="1059" alt="Screenshot 2022-11-23 at 14 43 16" src="https://user-images.githubusercontent.com/25876548/203506607-81c5f8fb-eafd-481e-94f5-15e7d992aa34.png">

sobitneupane commented:
I can see a lot of "(edited)" message out of nowhere. But could not reproduce the issue where report welcome message appears in the middle.

https://user-images.githubusercontent.com/25876548/203713079-998d422d-1cb5-4e38-bcde-8dc8887bd274.mov

[onyxdata_reportAction.txt](https://github.com/Expensify/App/files/10081459/onyxdata_reportAction.txt)

roryabraham committed:
Merge branch 'main' into Rory-SortReportActionsByCreated

roryabraham committed:
Remove actionName from sorting criteria and tests

roryabraham commented:
This PR has been substantially simplified to account only for `created` and `reportActionID`, and impose no sorting order based on `actionName`. This was an over-engineered consideration since same-millisecond reportActions are an edge case anyway, and I couldn't reproduce any unexpected behavior on this branch.

cead22 commented on src/libs/ReportActionsUtils.js:
```diff
@@ -28,6 +28,30 @@ Onyx.connect({
+ * @param {Boolean} [shouldInvertSortingOrder]
+ * @returns {Array}
+ */
+function sortReportActions(reportActions, shouldInvertSortingOrder = false) {
```
- Out of curiosity, why did you change the function name? I think `getSortedReportActions` is better and it doesn't sound like it update the report in place (by ref), like the call `sortReportActions(iouActions);` suggests
- What do you think of rename `shouldInvertSortingOrder` to be `ascending`? "should invert" doesn't say anything about the actual order so "ascending" or "descending" are better
- Can we make the `shouldInvertSortingOrder` param mandatory? There's 5 calls to this function, 3 without passing the param, and two passing true, so this isn't really a default and I think explicit params are better

roryabraham replied:
> Out of curiosity, why did you change the function name? I think getSortedReportActions is better and it doesn't sound like it update the report in place (by ref), like the call sortReportActions(iouActions); suggests 

It's because unless I'm mistaken this function technically _does_ update the report in place by ref.

cead22 replied:
That makes sense. That said, I think we should keep the old name and always use the return value of the array instead of relying on the in-place modification (we're currently doing a mix of the two, which I don't think we should do)

roryabraham replied:
Updated the name and consistently use the return value instead of inline modification.

roryabraham replied:
> What do you think of rename shouldInvertSortingOrder to be ascending? "should invert" doesn't say anything about the actual order so "ascending" or "descending" are better

Good point about "should invert" not saying anything about the actual order, but in NewDot all boolean variables should start with `is`, `should`, `has`, or something similar. So I updated this to `shouldSortInDescendingOrder`

roryabraham replied:
> Can we make the shouldInvertSortingOrder param mandatory? There's 5 calls to this function, 3 without passing the param, and two passing true, so this isn't really a default and I think explicit params are better

I disagree about this not being a default, because the places where we pass `true` and sort in descending order are for a component called `InvertedFlatList`, where the sorting order needs to be inverted. At least that's how I reason about the default value of this parameter.

cead22 commented on src/libs/ReportActionsUtils.js:
```diff
@@ -28,6 +28,30 @@ Onyx.connect({
+ */
+function sortReportActions(reportActions, shouldInvertSortingOrder = false) {
+    if (!_.isArray(reportActions)) {
+        throw new Error(`ReportActionsUtils::sortReportActions requires an array, received ${typeof reportActions}`);
```
```suggestion
        throw new Error(`ReportActionsUtils.sortReportActions requires an array, received ${typeof reportActions}`);
```
NAB, but I didn't see any other place where we use `::` like in PHP

roryabraham replied:
Updated

cead22 commented on src/libs/ReportActionsUtils.js:
```diff
@@ -28,6 +28,30 @@ Onyx.connect({
 });
 
+/**
+ * Sort an array of reportActions by their created timestamp,
+ * using actionName as a secondary sorting parameter,
+ * finally falling back on reportActionID in case multiple reportActions of the same actionName are created in the same millisecond.
```
```suggestion
 * Sort an array of reportActions by their created timestamp first, and reportActionID second
 * This gives us a stable order even in the case of multiple reportActions created on the same millisecond
```

roryabraham replied:
Updated

cead22 commented on src/libs/ReportActionsUtils.js:
```diff
@@ -38,37 +62,20 @@ function isDeletedAction(reportAction) {
+        return null;
+    }
+
+    sortReportActions(iouActions);
+    return _.last(iouActions).reportActionID;
```
```suggestion
    return _.last(sortReportActions(iouActions)).reportActionID;
```
Use the return value of `sortReportActions(iouActions);`

roryabraham replied:
Good call, done

cead22 commented on src/pages/home/report/ReportActionsList.js:
```diff
@@ -30,13 +30,7 @@ const propTypes = {
     /** Sorted actions prepared for display */
-    sortedReportActions: PropTypes.arrayOf(PropTypes.shape({
-        /** Index of the action in the array */
-        index: PropTypes.number,
```
`index` isn't needed anymore? I see it used in `renderItem` but maybe the update `item: reportAction,` takes care of this? Just curious

roryabraham replied:
Yeah, this isn't needed anymore. The `index` you see in `renderItem` comes from React Native itself.

cead22 replied:
Thanks for the tip

roryabraham committed:
Address review comments

roryabraham committed:
Rename shouldInvertSortingOrder to shouldSortInDescendingOrder

roryabraham committed:
Rename sortReportActions to getSortedReportActions

cead22 commented:
I need to retest, but the code looks good to me

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

cead22 reviewed (dismissed):
Approving to signal the code looks good, even though we still have to complete the reviewer checklist before merging

mountiny commented:
Ok, testing now to push this refactor forwards 🚀

mountiny commented:
I have updated Carlos' checklist https://github.com/Expensify/App/pull/12626#issuecomment-1329951556

There was on eissue I saw on iOs where the messages wwre initially not in correct order. I am on slow wifi and I think it influence the ngrok setup I have and so I think that was the problem also that we might need to update the optimistic actions too.

I was not able to repo on web so I can see the logs and onyx data. I don think this would be a blocker for the project. I would say this will be fixed with other PR and updating the optimistic actions

Web - I have encountered some issues with requesting money when offline, but I think that was not related to this PR. The error got cleared and after refresh all was fine, this might be just related to the other issue with how we process the write/read requests


https://user-images.githubusercontent.com/36083550/204496322-6ccf1ac4-d162-41d6-8b95-4ff841794324.mp4


Android


https://user-images.githubusercontent.com/36083550/204496863-55a709a2-5dbc-49bc-90f7-001450a40608.mp4

iOS

https://user-images.githubusercontent.com/36083550/204498933-12381dfd-e0e9-4a5c-a78a-05a4e3261ba4.mp4

mWeb Chrome

https://user-images.githubusercontent.com/36083550/204499267-dd7da7f2-6420-4c0b-aab0-b09e66cdc299.mp4

mountiny commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

</details>

sobitneupane commented:
@mounity I am also testing this PR. Should I continue?

mountiny commented:
> I am also testing this PR. Should I continue?

@sobitneupane if you started, go ahead 🙇

mountiny commented:
@sobitneupane you getting this even on new accounts or does this need to be old account?

sobitneupane commented:
@mountiny I am getting this on new account as well.


https://user-images.githubusercontent.com/25876548/204508345-7d46d6a0-43de-42f7-809d-e655992d6b34.mov

mountiny commented:
Thanks for confirming

roryabraham committed:
Merge branch 'main' into Rory-SortReportActionsByCreated

roryabraham committed:
Move function back to where it was to reduce diff

roryabraham committed:
Add filtering back to reportActions

roryabraham committed:
Filter out any unsupported action types

roryabraham committed:
Extract filter function to ReportActionsUtils

roryabraham committed:
Add tests for filter function

roryabraham commented:
Thanks for the testing and reproduction steps, @sobitneupane. I removed the filtering from the sorting function and forgot to add it back in a separate function. This PR has been updated and is ready for another review. I also added more unit tests to cover the latest changes.

cead22 commented:
- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.


@roryabraham can you update the manual tests to cover the cases that were raised in the review, and that led to the most recent updates? I think it's a good idea to manually test to make sure these are fixed (even if the auto tests cover the functionality). Once that's done I can test everywhere

roryabraham commented:
Updated the manual test + QA steps

roryabraham committed:
Fix automated test

cead22 commented:
Also @roryabraham I'm not sure how strict we are with screenshots, but the screenshots for some platforms don't show all the functionality being tested, do we require that that's shown either in videos or screenshot? I imagine we should, since otherwise the screenshot/video can be of anything and doesn't really show the tests was done

roryabraham committed:
Simplify boolean

roryabraham commented:
Personally I feel like this set of screenshots provides enough evidence to show that the sorting order of reportActions is as it should be. You're right that not every scenario is showcased on every platform, but all platforms and scenarios are shown. I also think that the automated tests I've provided should help provide additional confidence that this PR functions as expected.

cead22 commented:
Started testing and managed to test on web, but had trouble with iOS ([slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1669773091495239?thread_ts=1669772917.740239&cid=C01GTK53T8Q)). @mountiny if you can resume testing for the other platforms that'd be awesome to continue moving this forward. Otherwise I'll resume tomorrow

cead22 commented:
> Started testing and managed to test on web, but had trouble with iOS ([slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1669773091495239?thread_ts=1669772917.740239&cid=C01GTK53T8Q)). @mountiny if you can resume testing for the other platforms that'd be awesome to continue moving this forward. Otherwise I'll resume tomorrow

@sobitneupane if you can help us with this, that'd be great! Sorry for not tagging you in the first post, I'm still getting used to the processes

sobitneupane commented:
@roryabraham When I go offline and send messages, those messages shuffle few times before displaying in correct order.
 

https://user-images.githubusercontent.com/25876548/204745091-8c41191c-98dc-4c57-b8c4-905c7307e78f.mov

mountiny commented:
> @roryabraham When I go offline, send few messages and go online, those messages shuffle few times before displaying in correct order.

I had this same issue I think on iOS and I was not able to see the logs/ onyx state there. 

@roryabraham I would wait with my tessting before we decide how to fix this or if fix this in this PR. The created timestamp should not change when the report action becomes online but it seems from the video the messages were becoming "online" in the order they were sent, but the offline messages stayed as "older". Maybe that is something to look at in the logic.

cead22 commented:
I thought that was because Pusher messages aren't guaranteed to be received in the order they're sent

cead22 commented:
3 platforms down, 3 to go

mountiny commented:
> I thought that was because Pusher messages aren't guaranteed to be received in the order they're sent

If that is the case we are fine. 

I can help with the testing tomorrow morning if you wont get to ti, also @sobitneupane if you would be able to get to all platforms

sobitneupane commented:
Not only on offline mode, I can notice the shuffling issue online as well.

https://user-images.githubusercontent.com/25876548/204978891-50be539a-cc89-410e-a2f2-0360c7354b76.mov

sobitneupane commented:
On android, message sent offline, appear twice.

https://user-images.githubusercontent.com/25876548/204980197-14d7c11e-1a23-4d6e-a999-8322e256679f.mov

sobitneupane commented:
### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

</details>

mountiny reviewed (commented):
I have also experiences the reshuffling while being online which is odd 
<img width="1101" alt="image" src="https://user-images.githubusercontent.com/36083550/205060663-1513b951-a3ff-41ef-9f97-1ef494a56497.png">

this is just picture but the order of messages sent was 1, 2, 3, 4, 5 as you would expect.

I agree in the slack thread that if this is only in Offline we can do nothing for now, the end state is fine, but this should not happen when being online.

My reproduction steps have been simple, create a new chat and sent the messages in order and then it started to reshuffle, in the end it settles to the correct order but it is not great UX

mountiny commented:
Not sure what is going on but getting this behaviour, things are just really slow but it might be my VM, however, it also could be something else since Sobit experiences the same issue and he points to staging/prod.
https://user-images.githubusercontent.com/36083550/205063221-a3da88af-8c41-474c-91ce-d7851dcae853.mp4

mountiny reviewed (approved):
Alright, I have confirmed this exact behaviour is on staging/production as well 😬 try to type and send messages in quick succession and you will see they will shuffle and in general the requests wait for server response for a long time.

I feel like this must be something we probably already have reported somewhere but I will report in Slack just in case.

I have checked on the remaining boxes in checklist and I am approving as the issues I have found and Sobit found are on main and staging too so it is not related to this PR.

@roryabraham @cead22 Any reason why not to merge this?

mountiny commented:
The failing jobs are internal builds for testing because Rory is a mobile deployer, otherwise everything else is passing so we can merge this.

cead22 commented:
> The failing jobs are internal builds for testing because Rory is a mobile deployer, otherwise everything else is passing so we can merge this.

Discussed with @AndrewGable to understand this better, and I agree

cead22 commented:
As stated above, the builds that are failing are because of a bug that we're currently working to fix

cead22 commented:
The reviewer checklist was completed
<img width="473" alt="image" src="https://user-images.githubusercontent.com/165133/205125149-b07be16f-75fc-4678-bb63-5e5490415f0b.png">

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3604247225) to staging by @cead22 in version: 1.2.36-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3644465624) to production by @francoisl in version: 1.2.36-4 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

sobitneupane commented:
I am C+ in this PR. Can anyone please help me with the payment.

cc: @mountiny @cead22

mountiny commented:
Created issue here @sobitneupane https://github.com/Expensify/App/issues/13814

sobitneupane commented:
Thanks @mountiny. I have other 4 issues where payment are due. I have commented on respective issues as those issues are from Expensify/App. Should I link them as well?
